kernel/src/expressions/scalars.rs

```diff
@@ -66,9 +61,26 @@ pub struct ArrayData {
 }
 
 impl ArrayData {
-    pub fn new(tpe: ArrayType, elements: impl IntoIterator<Item = impl Into<Scalar>>) -> Self {
-        let elements = elements.into_iter().map(Into::into).collect();
-        Self { tpe, elements }
+    pub fn try_new(
+        tpe: ArrayType,
+        elements: impl IntoIterator<Item = impl Into<Scalar>>,
+    ) -> DeltaResult<Self> {
+        let elements = elements
+            .into_iter()
+            .map(Into::into)
+            .map(|v| {
```

Comment:
```suggestion
            .map(|v| {
                let v = v.into();
```
(there's no way two map calls is cheaper or simpler)